src/CleanArchitecture.Web/Startup.cs

```diff
@@ -58,14 +51,22 @@ private static IServiceProvider BuildDependencyInjectionProvider(IServiceCollect
             // Populate the container using the service collection
             builder.Populate(services);
 
-            // TODO: Add Registry Classes to eliminate reference to Infrastructure
+            // Reference all CleanArchitecture assemblies for AutoFac.
             Assembly webAssembly = Assembly.GetExecutingAssembly();
             Assembly coreAssembly = Assembly.GetAssembly(typeof(BaseEntity));
-            Assembly infrastructureAssembly = Assembly.GetAssembly(typeof(EfRepository)); // TODO: Move to Infrastucture Registry
+
+            string location = AppDomain.CurrentDomain.BaseDirectory;
+            Assembly infrastructureAssembly = GetInfrastructureAssembly(location);
             builder.RegisterAssemblyTypes(webAssembly, coreAssembly, infrastructureAssembly).AsImplementedInterfaces();
```

Comment:
Somehow it's possible to add functionality that will register additional interfaces that have custom classes? I mean autofac auto-scanning is able to register interfaces that class have different names and you don't need to manually do this?
Manually you can do something like [that](https://github.com/Xeinaemm/SpaTemplate/blob/c92853e33e48c583a6d4475ac37876c45a4a538c/src/SpaTemplate.Infrastructure.Core/Helpers/StartupExtensions.cs#L82), but at this moment you are splitting functionality into two places and it can be misleading.

Reply:
In other words, "put all custom registrations here, autofac will handle this for you".

Reply:
Another question is, what's about dynamic loading of other subprojects from Infrastructure, for example, we want to split ORM logic into **CleanArchitecture.Infrastructure.Orm**. And question it's we will add this functionality now when we are changing this?

Reply:
Right now there are no custom registrations that need to be made. But to register interfaces whose concrete classes live in Infrastructure, you can create a new Module in the Infrastructure project (or any other project where modules are registered (and as of right now, just Infrastructure is)).

An example module in this PR would be the `DatabaseModule`, where we register `AppDbContext` in there instead of in `Startup.cs` (since we no longer have an Infrastructure reference in Web. You could create your own Module (or just add more to `DatabaseModule`) to add registrations. A **Module** in Autofac is similar to a **Registry** in StructureMap.

Here's a sample of what could be made. `builder.RegisterAssemblyModules(infrastructureAssembly);` in `Startup.cs` will automatically scoop it in, if this class is in the Infrastructure project.

    public class InfrastructureModule : Module
    {
        protected override void Load(ContainerBuilder builder)
        {
            builder.RegisterType<ConcreteThing>().As<IThing>();
        }
    }

If you have registrations that are available in the Web project (i.e. can be put in `Startup.cs`), you can add them directly after the assembly registration lines.

    builder.RegisterAssemblyTypes(webAssembly, coreAssembly, infrastructureAssembly).AsImplementedInterfaces();
    builder.RegisterAssemblyModules(infrastructureAssembly);

    // Add custom registrations here. Or in modules in their respective assemblies.
    builder.RegisterType<ConcreteThing>().As<IThing>();

Reply:
To answer your question about loading of other dynamic projects, it would just be a matter of repeating the same steps demonstrated here to include the Infrastructure project. Adding the post-build command to copy the DLL to the Web project's output directory, then modify the code in `Startup.cs`.

    // Make "GetOtherAssembly" grab the full file name to the other assembly.
    Assembly otherAssembly = GetOtherAssembly(location);

    // Add your other assembly to the list of registered assembly types.
    builder.RegisterAssemblyTypes(webAssembly, coreAssembly, infrastructureAssembly, otherAssembly).AsImplementedInterfaces();

    // If you have a custom Module in your other assembly, add it here too.
    builder.RegisterAssemblyModules(infrastructureAssembly, otherAssembly);